### DISCUSSION TITLE
Why is VertexFormat not passable as usize ?

### DISCUSSION THREAD

**Discussion by erwanvivien** [gfx-rs/wgpu#3341] [General]
When using a VertexBufferLayout, we are forced to use `VertexFormat`, why can't we pass an abitrary number of bytes and let the shader take it as is ?

```rust
wgpu::VertexAttribute {
    offset: mem::size_of::<[f32; 3]>() as wgpu::BufferAddress,
    shader_location: 6,
    format: wgpu::VertexFormat::Float32x4,
    // format: mem::size_of::<f32>() * 16,
},
```

Wouldn't this be possible and portable ?

Edit: \
The usecase being passing a Mat4x4<f32> (to wgsl)

**Comment by grovesNL** [comment]
The graphics APIs used by wgpu's backends require the vertex attribute formats to be provided, so wgpu/WebGPU also require it (these can't always be derived from the shader later when the pipeline is used).

To pass a matrix you can use multiple vertex attribute locations.

**Reply by erwanvivien** [reply]
Why is this format needed ? If we send it as a byte blob, the shader can't interpret it or the inner steps ?

**Reply by grovesNL** [reply]
Yeah this step is all about describing the format of the byte blob. The graphics APIs all require this to be provided in advance, so wgpu can't skip this step.

The actual vertex buffer is sent as a byte blob.

**Reply by erwanvivien** [reply]
Alright, kinda sad about that, but I understand why is it necessary :) Thanks
